evals/elsuite/tools/eval.py

```diff
@@ -0,0 +1,85 @@
+import json
+import logging
+from typing import Any, List
+from urllib.parse import parse_qs, urlparse
+
+from datasets import load_dataset
+from pydantic import BaseModel
+
+import evals
+import evals.metrics
+from evals.api import CompletionFn
+from evals.record import RecorderBase
+
+logger = logging.getLogger(__name__)
+DEFAULT_SAMPLE_RATE = 16000
+
+
+class Sample(BaseModel):
+    messages: List[dict[str, Any]]
+    functions: List[dict]
+
+
+class ToolsTask(evals.Eval):
+    def __init__(
+        self,
+        completion_fns: list[CompletionFn],
+        dataset: str,
+        temperature: float = 0.7,
+        max_tokens: int = 4096,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(completion_fns, *args, **kwargs)
+        assert len(completion_fns) == 1, "Audio tasks only support one completion fn"
+        self.dataset = dataset
+        self.temperature = temperature
+        self.max_tokens = max_tokens
+
+    def load_sample(self, row) -> list[Sample]:
+        messages = row["messages"]
+        last_tool_call_id = None
+        for m in messages:
+            if m["tool_calls"]:
+                last_tool_call_id = m["tool_calls"][-1]["id"]
+            elif m.get("tool_call_id") is not None and last_tool_call_id is not None:
+                m["tool_call_id"] = last_tool_call_id
+                last_tool_call_id = None
+
+        return Sample(messages=messages, functions=json.loads(row["functions"]))
+
+    def eval_sample(self, sample: Sample, rng):
+        assert isinstance(sample, Sample)
+        prompt = sample.messages[:2]
+        assistant_msg = sample.messages[2]
+        try:
+            result = self.completion_fn(
+                prompt=prompt,
+                temperature=self.temperature,
+                max_tokens=self.max_tokens,
+                tools=[{"type": "function", "function": f} for f in sample.functions],
+            )
+            print("result", vars(result))
+            sampled = result.get_completions()[0]
+        except Exception as e:
+            logging.info("Sampling failed!")
+            logging.info(sample)
+            logging.info(f"Prompt: {prompt}")
+            logging.info(f"Error: {str(e)}")
+            sampled = "ERROR: " + str(e)
+        match = sampled == assistant_msg["content"]
+        evals.record.record_match(match, expected=assistant_msg["content"], sampled=sampled)
+        return match
+
+    def run(self, recorder: RecorderBase):
+        samples = self._get_dataset()
+        self.eval_all_samples(recorder, samples)
+        events = recorder.get_metrics()
+        return {"accuracy": evals.metrics.get_accuracy(events)}
```

Comment:
Did you say you found a metric for this as well? Otherwise I can use a simple check like [Gorilla](https://gorilla.cs.berkeley.edu/blogs/8_berkeley_function_calling_leaderboard.html#metrics). I'm thinking of creating two metrics: one that is 0/1, the other you can get partial points.

<img width="850" alt="image" src="https://github.com/user-attachments/assets/59e75e61-67c0-4b16-91e0-fd3500b5648f">

Reply:
Yeah, I was thinking one could just start with a dict compare on the params and have that be 0/1.